File: src/Ruby/coins.rb

# given an array of sorted coins of different values, what combination of their values will equal the sum in the least amount of coins...

def coins(coins, sum)
  combinations = Array.new(sum + 1){0}

  coins.each do |coin|
    (sum + 1).times do |idx|
      if coin <= idx
        combinations[idx] = combinations[idx - coin] + 1
      end
    end
  end

  combinations.last
end

File: src/JavaScript/coins.js
'use strict'

function coins(coins, sum){
	let combos = new Array(sum + 1).fill(0);

	for (let x = 0; x < coins.length; x++){
		let coin = coins[x];
		for (let i = 1; i < combos.length; i++){
			if (coin <= i){
				combos[i] = combos[i - coin] + 1
			}
		}
	}
	return combos
}

console.log(coins([1, 2, 5], 10));
